### PR TITLE
Implemented save loading error code logger

### PR DESCRIPTION
**Pull request by alphanin9**
This was originally also meant to have a check for what game system failed to load, but it turns out it's a bit complicated with async and all

### CODE REVIEW

File: src/dll/Hooks/ReportErrorCode.cpp
Comment (WopsS) [inline src/dll/Hooks/ReportErrorCode.cpp:1]: The name is a bit ambiguous, it doesn't really tell what system / class it is hooking.

File: src/dll/Hooks/ReportErrorCode.cpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+namespace
+{
+bool isAttached = false;
+
+void _State_SessionActive_ReportErrorCode(uintptr_t);
+Hook<decltype(&_State_SessionActive_ReportErrorCode)> State_SessionActive_ReportErrorCode(
+    0x7FA31576, &_State_SessionActive_ReportErrorCode);
+
+void _State_SessionActive_ReportErrorCode(uintptr_t aThis)
```
Comment (WopsS) [inline src/dll/Hooks/ReportErrorCode.cpp:15]: Same as the file's comment, the name is a bit ambiguous. I would prefer something like `CLASS/SYSTEM_FUNCNAME`.

File: src/dll/Hooks/ReportErrorCode.cpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+{
+    auto errorCode = *reinterpret_cast<uint32_t*>(aThis + 0x88u);
+
+    if (errorCode > 0u)
+    {
+        static const auto type = RED4ext::CRTTISystem::Get()->GetEnum("gsmStateError");
+
+        spdlog::error("=======");
+        spdlog::error("Session error, error code: {}", type->hashList[errorCode].ToString());
```
Comment (WopsS) [inline src/dll/Hooks/ReportErrorCode.cpp:24]: I assume `gsm` is GameStateManager? If so I would suggest a message like:

```suggestion
        spdlog::error("A game session error occurred. Error: {} ({})", type->hashList[errorCode].ToString(), errorCode);
```

**Review by WopsS — CHANGES_REQUESTED** [review]
Thanks for the PR! I have a few change requests.

File: src/dll/Hooks/ReportErrorCode.cpp (hunk 2 above)
Comment (alphanin9) [reply src/dll/Hooks/ReportErrorCode.cpp]: Likely GameStateMachine, as psiberx said

File: src/dll/Hooks/SessionActive_ReportErrorCode.cpp [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        spdlog::error("A game session error occurred. Error code: {} ({}).", type->hashList[errorCode].ToString(),
+                      errorCode);
+        spdlog::error("=======");
+    }
+
+    return State_SessionActive_ReportErrorCode(aThis);
+}
+} // namespace
+
+bool Hooks::SessionActive_ReportErrorCode::Attach()
```
Comment (WopsS) [inline src/dll/Hooks/SessionActive_ReportErrorCode.cpp:33]:
```suggestion
bool Hooks::gsmState_SessionActive::Attach()
```

File: src/dll/Hooks/SessionActive_ReportErrorCode.cpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    else
+    {
+        spdlog::trace("The hook for the game session error code reporter was attached");
+    }
+
+    isAttached = result == NO_ERROR;
+    return isAttached;
+}
+
+bool Hooks::SessionActive_ReportErrorCode::Detach()
```
Comment (WopsS) [inline src/dll/Hooks/SessionActive_ReportErrorCode.cpp:53]:
```suggestion
bool Hooks::gsmState_SessionActive::Detach()
```

**Review by WopsS — CHANGES_REQUESTED** [review]

**Review by WopsS — APPROVED** [review]
Thanks!

**Pull request merged**